src/wallet/walletdb.cpp

```diff
@@ -634,13 +634,32 @@ static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch,
             uint32_t index = 0;
             if (keyMeta.hdKeypath != "s" && keyMeta.hdKeypath != "m") {
                 std::vector<uint32_t> path;
-                if (keyMeta.has_key_origin) {
-                    // We have a key origin, so pull it from its path vector
-                    path = keyMeta.key_origin.path;
-                } else {
-                    // No key origin, have to parse the string
-                    if (!ParseHDKeypath(keyMeta.hdKeypath, path)) {
-                        strErr = "Error reading wallet database: keymeta with invalid HD keypath";
+                // Always parse the path string
+                if (!ParseHDKeypath(keyMeta.hdKeypath, path)) {
+                    strErr = "Error reading wallet database: keymeta with invalid HD keypath";
+                    return DBErrors::NONCRITICAL_ERROR;
+                }
+                // If there is a key origin, make sure that path matches
+                if (keyMeta.has_key_origin && path != keyMeta.key_origin.path) {
+                    // Detect if this metadata has a bad derivation path from a bug in inactive hd key derivation that was present in 0.21 and 22.x
+                    // See https://github.com/bitcoin/bitcoin/issues/29109
+                    // Markers for bad derivation:
+                    // - 6 indexes
+                    // - path[5] == path[2] + 1
+                    // - path[0:2] == path[3:5]
```

Comment:
Isn’t the line 648 in contradiction with line 649?

Looking at https://github.com/bitcoin/bitcoin/issues/29109, I’m thinking that the indices are referring to the elements separated by "/" after "m" of the hdkeypath variable here.

```diff
-[..] hdkeypath=m/0'/0'/299'/0'/0'/300'
+[..] hdkeypath=m/0'/0'/300'
```

I am wondering whether this should be:

```suggestion
                    // - path[5] == path[2] + 1
                    // - path[0:1] == path[3:4]
```

Reply:
Oh, are these ranges meant to be exclusive of the second parameter?

Reply:
Changed as suggested. Also added a bit more of an explainer to the comment that should make it clearer.